### PR TITLE
Add more stats to the MySQLCollector

### PR DESCRIPTION
**Pull request by tsheasha**
Now the collector produces on demand stats about:
    * Innodb rollback segments
    * Processlist idle threads and long queries
    * Temp table sizes on disk

### CODE REVIEW

**Comment by gnufied** [comment]
Looks like - this collector is leaking created cursors. It is probably out of scope for this pull request, but we don't want this thing to cause problems in production usage.

**Comment by tsheasha** [comment]
@gnufied nope cursors are not being leaked, The connection is only closed when either the collection fails or the collector dies. The cursor need not to be closed otherwise since we do a fetchall which gets all data in memory and the next time we gather stats we create a new cursor and this one is garbage collected. We do not want to close and recreate a connection for each connection that is why we have only one connection to the db.

**Comment by gnufied** [comment]
well, I was just going by the documentation and it says:

```
You MUST retrieve the entire result set and
    close() the cursor before additional queries can be peformed on
    the connection.
```

I could be wrong and further investigation will require going looking into implementation details of mysqldb python adapter and its C bindings.

**Comment by tsheasha** [comment]
@gnufied I added the cursor.close() for readability purposes. I am not sure how that statement from the docs affects us since we are using a new cursor per time anyways so that statement does not affect our situation. In practice this collector is being used in production and doing multiple queries per connection, just using different cursors. makes sense?

**Comment by gnufied** [comment]
yeah like I said. I could be wrong and it is probably fine leaving it as it is. But it seems like a no-brainer to err on the side of caution.

File: src/diamond/collector.py [hunk 1]
```diff
@@ -317,6 +317,7 @@ def publish_metric(self, metric):
             metric.dimensions or {}
         )
         payloadStr = "%s\n" % json.dumps(payload)
+        return
         success = False
```
Comment (gnufied) [inline src/diamond/collector.py]: Is this intentional?

Comment (tsheasha) [inline src/diamond/collector.py]: fixed

**Comment by gnufied** [comment]
:shipit: I would have loved to have tests but existing collector didn't had many tests to begin with.

File: src/diamond/collectors/mysqlstat/mysqlstat.py [hunk 2]
```diff
@@ -285,9 +304,12 @@ def get_db_stats(self, query):
 
         try:
             cursor.execute(query)
-            return cursor.fetchall()
+            result = cursor.fetchall()
+            cursor.close()
+            return result
         except MySQLError, e:
             self.log.error('MySQLCollector could not get db stats', e)
+            cursor.close()
```
Comment (tiras-j) [inline src/diamond/collectors/mysqlstat/mysqlstat.py]: Could put the close in a finally: here, just saves a line but also makes it clear that it should always occur.

Comment (tsheasha) [inline src/diamond/collectors/mysqlstat/mysqlstat.py]: done!

**Comment by tiras-j** [comment]
Shipit :)

File: src/diamond/collectors/mysqlstat/mysqlstat.py [hunk 3]
```diff
@@ -413,6 +505,12 @@ def _publish_stats(self, nickname, metrics):
             for metric_name in metrics[key]:
                 metric_value = metrics[key][metric_name]
 
+                if type(metric_value) is dict:
```
Comment (poros) [inline src/diamond/collectors/mysqlstat/mysqlstat.py]: it should be alright in this case, but this is kinda unsafe, please switch to `isinstance`

Comment (tsheasha) [inline src/diamond/collectors/mysqlstat/mysqlstat.py]: true I shouldn't have missed that, I might have gotten carried away by the other type check for float a couple of lines down.

**Comment by poros** [comment]
+1

**Pull request merged**